test/cpp/jit/test_jit_cpp_api.cpp

```diff
@@ -0,0 +1,401 @@
+naming our API (e.g. obj.attr, accessors). For our publically exposed API the
+root type's name is torch::Value. It internally wraps in IValue, and manually
+exposes what we want to be in the public API. A torch::Type object similar wraps
+our TypePtr internal type heirarchy.
+
+Unlike Python we have statically typed and tagged values, so instead of exposing
+generic methods on Value, we instead have a limited number of subtypes that
+mirror our Type hierarchy, and expose the appropriate functions for these types
+(e.g. Dict, List, Object, Tuple). These types inherit from Value, the base type
+of the public API.
+
+The wrapper Type similarly wraps up our internal TypePtr objects.
+
+
+
+Private Type          Public Type
+----------------------------------
+IValue                torch::Value
+TypePtr               torch::Type
+ivalue::Object        torch::Object
+*/
+
+namespace Test {
+
+struct Dict;
+struct List;
+struct Tuple;
+struct Object;
+struct Value;
+struct ListType;
+struct DictType;
+struct TupleType;
+struct ObjectType;
+struct Module;
+
+struct Type {
+  // type singletons
+  static Type Int();
+  static Type Float();
+  static Type Tensor();
+
+  // ability to do comparisions
+  bool isSubtypeOf(const Type& rhs);
+
+  // destructuring functions, avoid exposing tags, etc.
+  // to make this more maintainable.
+  bool isListType() const;
+  ListType toListType() const;
+
+  // similar for List, Dict, Tuple, Object
+ private:
+  c10::TypePtr typePtr_;
+};
+
+struct Class : public Type {
+  // note: no creation API, this is purely for interacting with existing types
+
+  // constructor for an object of this class
+  template <typename... Args>
+  Object operator()(Args... args);
+};
+
+struct ListType : public Type {
+  ListType(Type value);
+  // accessors for subtypes
+  Type element() const;
+  Type value() const;
+};
+
+struct DictType : public Type {
+  DictType(Type key, Type value);
+  // accessors for subtypes
+  Type key() const;
+  Type value() const;
+};
+
+struct TupleType : public Type {
+  TupleType(std::vector<Type> values);
+  TupleType(std::vector<std::string> field_names, std::vector<Type> values);
+  // accessors for subtypes
+  const std::vector<Type>& elements() const;
+};
+
+// All the methods that go on a Value get defined here. This makes it easy for
+// accessor objects, used to represent the LHS of assignments to also have these
+// methods through inheritance. This design is adapted from pybind11
+using jit::TypePtr;
+template <typename Derived>
+struct ValueAPI {
+  // example to, but in the real thing
+  // we should wrap by hand, so we have control over what is publically exposed
+  template <typename T>
+  T to() const {
+    derived().template to<T>();
+  }
+  // example accessors, implemented for Int only in prototype
+  bool isInt() const {
+    return derived().isInt();
+  }
+  bool toInt() const {
+    return derived().toInt();
```

Comment:
`toInt()` today throws an internal error when the `IValue` is not an int. I assume we'll want to convert that into a more descriptive user-error that includes the type tag? How should we do this without doubling the size of the IValue API?

Reply:
We can consider having both IValue and Object inherit from the same ValueAPI. ValueAPI is already CRTP so we can have it use methods of the base class to determine how to do the error checking. For IValue it will produce a internal error and for Value it would produce a user facing error.  This would require a further refactoring of things.  The easy part here is to define what we want the public API to be. The hard part is how we maintain both a public and private API that do not get wildly duplicative and inconsistent.  I am not sure how to do this yet.